mtt/src/mtt_types.rs

```diff
@@ -0,0 +1,616 @@
+    Non64MBL2Entry,
+    /// The caller specified a operation on a 4K page, but the L2 entry for the physical
+    /// address doesn't map a 4K page.
+    Non4KL2Entry,
+    /// The caller specified an invalid physical address (too large for example)
+    InvalidAddress,
+    /// The platform initialization hasn't been completed. The implementation is expected
+    /// to initialize pointers to the MTT L2 and L1 page pool using platform-FW tables.
+    PlatformInitNotCompleted,
+    /// The platform didn't allocate sufficient L1 page pool pages to map the specified address
+    InsufficientL1Pages,
+}
+
+/// Holds the result of a MTT operation.
+pub type Result<T> = core::result::Result<T, Error>;
+
+const MTT_L2_ENTRY_COUNT: usize = 1024 * 1024;
+const MTT_L1_ENTRY_COUNT: usize = 512;
+// Struct to represent a L1 page
+// Each entry represents 32 4K pages, which represents a total of 64MB of address space.
+// Each page uses a 2-bit encoding, with '00b as non-confidential and '01b as confidential ('11b is illegal).
+struct MttL1Page {
+    base: [u64; MTT_L1_ENTRY_COUNT],
+}
+
+// Encapsulates core platform functionality, including pointers to the platform MTT L2 and platform L1 page pool
+struct MttCore {
+    mtt_l2_base: &'static mut [u64],
+    mtt_l1_page_pool_base: &'static mut [MttL1Page],
+}
+
+static mut MTT_CORE: Once<MttCore> = Once::new();
+// Utility structure for for L2 manipulation
+struct MttHelper {
+    // Type of MTT L2 entry for the physical address in `phys_addr`
+    l2_entry: MttL2EntryType,
+    phys_addr: SupervisorPageAddr,
+}
+
+#[allow(dead_code)]
+impl MttCore {
+    /// Creates a new MttL1PagePool object.
+    /// # Safety -- the caller must ensure that:
+    /// 1: `l2_base` is a physically contiguous and reserved 8MB region allocated by platform-FW corresponding to the MTT L2.
+    /// 2: `l1_pool_base` is a valid pointer to a contiguous reserved region of `l1_pool_page_count` 4K-pages. Note that for computation purposes,
+    /// we assume that the platform memory is a contiguous range beginning at physical address 0 and ending at the top of physical memory. The caller must
+    /// allocate 16 4K pages in the pool for each 1GB of addressable space.
+    /// 4: `l2_base_addr` and `l1_pool_base_addr` must not be aliased elsewhere, and must remain allocated until a platform reset.
+    pub unsafe fn new(l2_base: u64, l1_pool_base: u64, l1_pool_page_count: u64) -> Self {
+        let mtt_l2_base = core::slice::from_raw_parts_mut(l2_base as *mut u64, MTT_L2_ENTRY_COUNT);
+        let mtt_l1_page_pool_base = core::slice::from_raw_parts_mut(
+            l1_pool_base as *mut MttL1Page,
+            l1_pool_page_count as usize,
+        );
+        Self {
+            mtt_l2_base,
+            mtt_l1_page_pool_base,
+        }
+    }
+
+    // Returns a static reference to instance if platform initialization has been completed
+    fn get() -> Result<Mutex<&'static mut MttCore>> {
```

Comment:
This probably has to return a MutexGuard to be useful, and every call should be a Guard from the _same_ global mutex.

This code mints a new mutex for each call so doesn't provide mutual exclusion.